### ISSUE TITLE
?win32 and ?win64 interpretation is inconsistent

### ISSUE THREAD

**Issue by GWRon** [bmx-ng/bcc#361]
There is the (vanilla/legacy) conditional `?win32` and there are 3 other NG introduced conditionals: `?win64 ?win32x86 ?win32x64`.

If you knew about win32 and win64 you would think the first one is for 32bit builds and the other one for 64bit builds. Yet the win32 variant is true for every Windows (so 32 and 64bit) while the 64bit one is only valid for the 64bit Windows builds.

This can lead to wrong code when misinterpreted.

For details see my reply in the syntaxbomb forums:
https://www.syntaxbomb.com/index.php/topic,5067.msg21505.html#msg21505

While `?win32` is there for compatibility reasons the `?win64` only exists for ... dunno. It is the same condition than with `?ẁin32x64` - so could get removed at all. This would lead to only the 3 conditionals left - and as there is no ("word-wise") counterpart for `?win32` it might be less inconsistent.

In short: if there is no reason to keep it, `?win64` could get removed.

**Comment by HurryStarfish** [comment]
The documentation on this is misleading/wrong too.
It says

> Name | Meaning
> -- | --
> Win32 | True if program is being compiled for 32 bit Windows operating system.
> Win32x86 | Same as above
> Win64 | True if program is being compiled for 64 bit Windows operating system.
> Win32x64 | Same as above

Yet, as you say, on my 64bit Windows 7, `Win32`, `Win64` and `Win32x64` are all true, only `Win32x86` isn't.
Technically the new ones are all redundant. `Win32x86` and `Win32x64` are the same as `Win32 And x86` and `Win32 And x64` respectively. But I agree `Win64` is particularly confusing.

**Comment by GWRon** [comment]
Another thing to consider: NG allows to concat multiple conditionals. So if backwards compatibility is not of an issue then you could also have something like

?windows and 64bit
print "64bit Windows"
?64bit
print "non-Windows 64bit OS"
?

**Comment by davecamp** [comment]
I suggest

?windows ' = 32 and 64 bits
?win32 ' = windows 32 bit only
?win64 ' = windows 64 bit only

arguably it may well break backwards compatibility but then it cleans things up going forwards.

**Comment by HurryStarfish** [comment]
In addition, breaking changes in most places - as long as you're aware of them - can at least be worked around by wrapping the code in question in `? BMXNG`/`? Not BMXNG` to allow for legacy/NG cross-compatible code. Breaking changes to the precompiler flags can't.

**Comment by markcwm** [comment]
I don't like col's suggestion, backward compatibility is more important here. I also don't like the suggestion to remove Win64 as it may now be in use somewhere, and it can't do much harm if the docs are clear on this, as Hurrystarfish points out we need a correction ie. "Win32" should be "..compiled for Windows.." and "win32x86" should be "..compiled for 32 bit Windows.." other than that the "Conditional Compiling" section looks okay to me.

**Comment by HurryStarfish** [comment]
Since, as @GWRon and @markcwm mentioned, undefined precompiler flags default to false and somone might already have used `Win64`, removal might also lead to the silent introduction of bugs into their code.
I'd suggest a "grace period" during which the flag would still be present, but bcc would output a warning stating that it is deprecated, will eventually be removed and shouldn't be used anymore.
